### PR TITLE
Add RO/RW span interfaces

### LINKED ISSUE #1326: Align span interfaces with the spec
Following the discussion [here](https://github.com/open-telemetry/opentelemetry-go/issues/1304#issuecomment-726206274), we should take another look at the way we use `export.SpanData` in `sdk/trace`.

The spec [mentions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#additional-span-interfaces) two types of interfaces to a span at the SDK level:

- A read-only span
- A read/writable span

Right now we use `export.SpanData` in the SDK as a "container" for any information about a span which may need to be read, for example for the sake of exporting a span. We store a pointer to a `SpanData` struct in a span's `data` field and we make a **copy** of the `SpanData` using `makeSpanData()` when we want to prevent modifications to an ended span:

https://github.com/open-telemetry/opentelemetry-go/blob/27aa1f601124468cde6c796935460435b6723700/sdk/trace/span.go#L52

https://github.com/open-telemetry/opentelemetry-go/blob/27aa1f601124468cde6c796935460435b6723700/sdk/trace/span.go#L146

We should see if there is a cleaner, Go-idiomatic way to express the two "behaviors" of the span inside the SDK and refactor accordingly.

### PR DESCRIPTION
**Pull request by johananl**
## General

This PR refactors the way we handle access to span data in the SDK.

Fixes #1326.

Reviewing this PR _commit by commit_ can be easier.

This PR touches code in the "critical path" of the SDK. I suggest we keep the following in mind when reviewing:

- We should ensure we aren't making performance worse by merging this.
- We should verify the set of methods under `ReadOnlySpan` make sense: Are we exposing everything which might need to be read? Are we exposing anything unnecessary?

## Rationale

The spec defines [multiple interfaces](https://github.com/open-telemetry/opentelemetry-specification/blob/6c8dab9d97fe4f0441f394864420dd5bebdbc46c/specification/trace/sdk.md#additional-span-interfaces) for interacting with spans. These interfaces make a clear distinction between the following:

- Places in the codebase where a span should only be written to.
- Places in the codebase where a span should only be read from.
- Places in the codebase where a span might have to be either read from or written to.

This change improves the way we control access to the various fields of a `span` struct in the SDK.

In addition, this change explicitly designates `export.SpanData` as a **read-only snapshot** of a span. Following this change, `export.SpanData` instances are no longer modified after creation and are used solely for exporting spans.

## Summary of changes

- Store all span data directly in the `span` struct.
- Add a `ReadOnlySpan` interface which matches the "readable span" concept from the spec.
- Add a `ReadWriteSpan` interface which matches the "read/write span" concept from the spec.
- Use the two interfaces above as appropriate instead of `export.SpanData`.

See the commit messages for more info about each change.

## Testing

The easiest way to test this PR is to run the following unit tests:

```
cd sdk
go test -v ./trace -run TestReadOnlySpan
go test -v ./trace -run TestReadWriteSpan
```

### CODE REVIEW

File: sdk/trace/attributesmap.go
Comment (Aneurysm9) [inline sdk/trace/attributesmap.go:71]: Should this pre-allocation of capacity for the `attributes` slice be retained?  We already know how large it is going to be and can potentially save some reallocation.

File: sdk/trace/span.go
Comment (Aneurysm9) [inline sdk/trace/span.go:216]: If setting the end time were moved up it would be necessary to also move up the `if !s.IsRecording()` check above (or remove it entirely, perhaps).

**Review by Aneurysm9 — APPROVED** [review]

File: sdk/trace/attributesmap.go
Comment (XSAM) [inline sdk/trace/attributesmap.go:72]:
```suggestion
// toKeyValue copies the attributesMap into a slice of label.KeyValue and
// returns it. If the map is empty, a nil is returned.
// TODO: Is it more efficient to return a pointer to the slice?
func (am *attributesMap) toKeyValue() []label.KeyValue {
	if am.evictList.Len() == 0 {
		return nil
	}
	
	var attributes []label.KeyValue
```

Nit

Comment (johananl) [reply sdk/trace/attributesmap.go:72]: Thanks! I actually wasn't aware nil was the zero value of a slice. Fixed.

File: sdk/trace/attributesmap.go
Comment (johananl) [reply sdk/trace/attributesmap.go:71]: Makes sense to me. Fixed.
For my own understanding, does this affect performance in a meaningful way?

File: sdk/trace/span.go
Comment (johananl) [reply sdk/trace/span.go:216]: Makes sense. I'll take a look at this. I'll try to check if storing the end time early has any impact at all.

File: sdk/trace/attributesmap.go
Comment (johananl) [reply sdk/trace/attributesmap.go:71]: I've added a benchmark for this function and pre-allocating like you've suggested is about 50% faster. Great catch!

Comment (johananl) [reply sdk/trace/attributesmap.go:71]: Without pre-allocation:

```
go test ./trace -bench BenchmarkAttributesMapToKeyValue
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
BenchmarkAttributesMapToKeyValue-8        224655              4562 ns/op
PASS
ok      go.opentelemetry.io/otel/sdk/trace      1.231s
```

With pre-allocation:

```
go test ./trace -bench BenchmarkAttributesMapToKeyValue
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
BenchmarkAttributesMapToKeyValue-8        534184              2142 ns/op
PASS
ok      go.opentelemetry.io/otel/sdk/trace      1.321s
```

**Comment by krnowak** [comment]
The question that pops up is: Would it be ok to rename `SpanData` to something like `SpanSnapshot`? I realize that java implementation is calling it `SpanData`, but `SpanSnapshot` actually describes that thing better.

**Comment by johananl** [comment]
> The question that pops up is: Would it be ok to rename `SpanData` to something like `SpanSnapshot`? I realize that java implementation is calling it `SpanData`, but `SpanSnapshot` actually describes that thing better.

I agree `SpanSnapshot` would make things less ambiguous. I'd let some more folks weigh in before changing this though.

**Comment by Aneurysm9** [comment]
> The question that pops up is: Would it be ok to rename `SpanData` to something like `SpanSnapshot`? I realize that java implementation is calling it `SpanData`, but `SpanSnapshot` actually describes that thing better.

I agree that `SpanSnapshot` more clearly articulates that this is a point-in-time snapshot of the data.  I'd also suggest changing `Dump()` to `Snapshot()` to more clearly align the method name with the type it produces.

File: sdk/trace/span.go
Comment (johananl) [reply sdk/trace/span.go:219]: I've tried sleeping inside `End()` before setting `s.endTime` and as expected this makes the span's duration "artificially" long. As far as I can tell we want to store the end timestamp ASAP because the thing we are tracing is the **user** operation, not the time it takes us to finish a span (right?).

>If setting the end time were moved up it would be necessary to also move up the if !s.IsRecording() check above (or remove it entirely, perhaps).

Actually @Aneurysm9 I don't have to touch `IsRecording()` because I only store the end time in a new variable at the top, and then set it in `span.endTime` only if we get to this part of the function (i.e. if we didn't return earlier).

I have another problem though: If `End()` gets called more than once, my current logic overwrites the original end time which shouldn't happen. I'm looking into this.

Comment (johananl) [reply sdk/trace/span.go:219]: Looks like we might be using `s.endOnce` unnecessarily because if `End()` is called a 2nd time, `IsRecording()` returns `false` and we correctly return without trying to overwrite the end time. I'll open a new PR to refactor `End()` as this is starting to be out of scope here IMO.

Comment (johananl) [reply sdk/trace/span.go]: >I'll open a new PR to refactor End() as this is starting to be out of scope here IMO.

...or not :see_no_evil: I don't have the right interfaces for refactoring properly until this is merged. Oh well... Let me see what I come up with.

Comment (johananl) [reply sdk/trace/span.go]: I've pushed a new commit which refactors `span.End()`. Looks like we can address the problem discussed here and also simplify the idempotence logic.

File: sdk/trace/span.go
Comment (jmacd) [inline sdk/trace/span.go:55]: (On this TODO: :+1:)

File: sdk/trace/span.go
Comment (jmacd) [inline sdk/trace/span.go:312]: I'd prefer if the exporters aren't going to have to take and release a whole bunch of Mutexes to read through a ReadOnly type. Can the ReadOnly accessors be simple accessors, and then could the ReadWrite accessors just take a lock and call through to the ReadOnly accessors? It seems that by the time you pass a span to an ordinary exporter, you no longer need the mutex. Is that possible?

**Review by jmacd — CHANGES_REQUESTED** [review]

File: sdk/trace/span.go
Comment (johananl) [reply sdk/trace/span.go:55]: @jmacd did you mean that it makes sense for `Event` to live under `export`?

File: sdk/trace/span.go
Comment (johananl) [reply sdk/trace/span.go:312]: Thanks for the feedback @jmacd.

When exporting a span we actually lock the mutex just once: The span processor calls `span.Dump()` (which I'll likely rename following the feedback above) which locks the mutex, takes a "snapshot" of the span's state and then releases the mutex.

Example:

https://github.com/johananl/opentelemetry-go/blob/e81b72935d5b1e0c9378906f2c7b2161062af36f/sdk/trace/simple_span_processor.go#L46-L53

The RO accessors are to be used when only a specific span field needs to be read.

Does that make sense to you?

**Comment by johananl** [comment]
Following the feedback above I've renamed `SpanData` to `SpanSnapshot` and `span.Dump()` to `span.Snapshot()`.

File: sdk/trace/span.go
Comment (MrAlias) [inline sdk/trace/span.go:55]: It does not make sense to me for it to live in the export package. I think it should be moved to the SDK.

File: sdk/trace/span.go
Comment (MrAlias) [inline sdk/trace/span.go:43]: I think based on the separation of concerns between the API and the SDK this redundancy makes sense. The readable parts of the span could be abstracted into an internal interface shared between the two parts, but it would couple these concerns and not provide a definite interface definition to the user.

If we omitted these methods here we there could be situations where the user would need to assert that a value of this interface also satisfied the API's span which would not be ideal.

I think having these defined here makes things explicit, enforces the the separation of concerns between the API and SDK, and provides a unified useful interface to the user.

The only issue I see is if the methods from the API and SDK start to differ in signature but keep the same name. I don't think there is a way around this given the requirement that the API and SDK be separate though.

**Review by MrAlias — COMMENTED** [review]
The `ReadOnlySpan` and `ReadWriteSpan` interfaces look good :+1:. I think the `Event` type can be moved to the SDK, but that's not blocking the interface changes.

Based on @jmacd's locking comment and the updated needed to the `ExportSpan` method, I think our overall approach to using these interfaces, specifically the `ReadOnlySpan` interface, needs to be updated to fully commit to them.

I agree with the name change to the `SpanData` to `SpanSnapshot`, but I do not think that struct belongs in the export package anymore. To conform to the specification and also actually provide a read-only interface (instead of just an agreed upon use of `SpanSnapshot`) the `ExportSpan` interface should be updated to accept the `ReadOnlySpan`. If that update is made it would make sense for the `SpanSnapshot` to be moved into the SDK itself as an unexported type. This type could continue to not have any locks (with it being unexported we can control field access internal to the package) and could be passed to the export pipeline as a `ReadOnlySpan`. The existing `span` type in the SDK can then remain one that provides concurrent safe methods to read and write the span fields and will snapshot those fields when exported.

I think this secondary step of updating the `SpanExport` interface and moving the `SpanData`/`SpanSnapshot` into the SDK should be a follow on PR leaving the addition of the `ReadOnlySpan` and `ReadWriteSpan` interfaces to the singular purpose of this PR.

@XSAM @Aneurysm9 @jmacd @johananl thoughts?

**Comment by johananl** [comment]
> The `ReadOnlySpan` and `ReadWriteSpan` interfaces look good . I think the `Event` type can be moved to the SDK, but that's not blocking the interface changes.
> 
> Based on @jmacd's locking comment and the updated needed to the `ExportSpan` method, I think our overall approach to using these interfaces, specifically the `ReadOnlySpan` interface, needs to be updated to fully commit to them.
> 
> I agree with the name change to the `SpanData` to `SpanSnapshot`, but I do not think that struct belongs in the export package anymore. To conform to the specification and also actually provide a read-only interface (instead of just an agreed upon use of `SpanSnapshot`) the `ExportSpan` interface should be updated to accept the `ReadOnlySpan`. If that update is made it would make sense for the `SpanSnapshot` to be moved into the SDK itself as an unexported type. This type could continue to not have any locks (with it being unexported we can control field access internal to the package) and could be passed to the export pipeline as a `ReadOnlySpan`. The existing `span` type in the SDK can then remain one that provides concurrent safe methods to read and write the span fields and will snapshot those fields when exported.
> 
> I think this secondary step of updating the `SpanExport` interface and moving the `SpanData`/`SpanSnapshot` into the SDK should be a follow on PR leaving the addition of the `ReadOnlySpan` and `ReadWriteSpan` interfaces to the singular purpose of this PR.
> 
> @XSAM @Aneurysm9 @jmacd @johananl thoughts?

@MrAlias your suggestions make sense to me. I'm happy to follow up either in this PR or open follow up issues. I'll probably start authoring the commits for that in the meantime and I can cherry-pick them into a new branch if necessary.

**Comment by johananl** [comment]
@MrAlias moving `Event` to `sdk/trace` creates an import cycle:

```
package go.opentelemetry.io/otel/example/basic
        imports go.opentelemetry.io/otel/exporters/stdout
        imports go.opentelemetry.io/otel/sdk/export/trace
        imports go.opentelemetry.io/otel/sdk/trace
        imports go.opentelemetry.io/otel/sdk/export/trace: import cycle not allowed
make: *** [Makefile:92: build] Error 1
```

I think this will be resolved if we move `SpanSnapshot` out of `sdk/export` as you suggested. I'll update.

**Comment by MrAlias** [comment]
> Should `sdk/export` depend on `sdk/trace` at all? I suspect not.

It should not.

> I'm happy to follow up either in this PR or open follow up issues.

I would recommend splitting into multiple PRs and I like your suggestion to track the second phase in its own issue. :+1:

**Comment by johananl** [comment]
> > Should `sdk/export` depend on `sdk/trace` at all? I suspect not.
> 
> It should not.

Thanks @MrAlias.
Are you sure about this? I feel like this needs some more consideration. Here is why:

I've looked at this again and commented [here](https://github.com/open-telemetry/opentelemetry-go/pull/1360#issuecomment-734407204) that at least in the Java library the "direction" of the dependency seems to be `sdk/export` -> `sdk/trace`.

I *think* this approach uses the *dependency inversion principle* which we seem to already be following and I think it could make a lot of sense to use this pattern in the Go library, too:

- A _high-level_ module should not depend on a _low-level_ module.
- Inter-module dependencies should be done using _interfaces_, not concrete types.

Translating this to our case:

- `sdk/trace` is arguably higher level than `sdk/export` (because it "consumes" things from `sdk/export` and wires them together with other entities), which means `sdk/trace` should not depend on `sdk/export`.
- `sdk/export` depends on interfaces exposed by `sdk/trace` such as `SpanProcessor` and `ReadOnlySpan`.
- `SpanProcessor` implementations should live inside `sdk/export` (we need to move them).

AFAICT the same pattern is already present in the relationship between the API and the SDK:

- The API defines interfaces (such as `Span`, `Tracer` and `TracerProvider`) and doesn't depend on the SDK.
- The SDK depends on the interfaces exposed by the API.

Given the above, my current feeling is that we should keep using this approach for `sdk/trace` and `sdk/export`.

In any case, right now we don't have a clear demarcation between `sdk/trace` and `sdk/export` so the interface between these two packages has to be refactored IMO.

Happy to discuss this in a meeting.

File: sdk/export/trace/trace.go
Comment (johananl) [reply sdk/export/trace/trace.go:43]: I'll resolve this since this is discussed in https://github.com/open-telemetry/opentelemetry-go/issues/1380 and in any case can't be done here due to the import cycle problem.

File: sdk/trace/span.go
Comment (johananl) [reply sdk/trace/span.go:55]: I'll resolve this since this is discussed in #1380 and in any case can't be done here due to the import cycle problem.

File: sdk/trace/span.go
Comment (johananl) [reply sdk/trace/span.go:43]: Thanks for the feedback :+1: This is how I see this, too.

**Comment by johananl** [comment]
Current status: We've moved the remaining discussions to #1380 in order to keep the scope of this PR narrow.

As far as I'm concerned this is ready for merging. There is one open concern from @jmacd [here](https://github.com/open-telemetry/opentelemetry-go/pull/1360#discussion_r529560938), however AFAICT the problem described there isn't actually the case so I'm not sure if there's anything to do there.

**Review by MrAlias — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

File: sdk/trace/span.go
Comment (MrAlias) [reply sdk/trace/span.go:309]: Going to resolve this. @jmacd feel free to unresolve if needed.

File: CHANGELOG.md
Comment (MrAlias) [inline CHANGELOG.md:27]: These edits need to be moved to the `Unreleased` section.

Comment (johananl) [reply CHANGELOG.md:27]: Oh! Didn't realize I was writing in the wrong place. Thanks :+1:

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
@jmacd I think all your concerns have been addressed, can you please take another look?

**Review by Aneurysm9 — APPROVED** [review]

**Review by jmacd — APPROVED** [review]

**Comment by jmacd** [comment]
Thanks @johananl

**Comment by MrAlias** [comment]
Holding off on including this in the v0.15.0 release so we can hope to address the changes for the `sdk/trace/export/trace` package all in one release.

**Review by XSAM — APPROVED** [review]

**Pull request merged**
